server/src/api/gitLab/gitLabIntegration.ts

```diff
@@ -0,0 +1,213 @@
+import { OAuth2 } from 'oauth';
+import { randomBytes } from 'crypto';
+
+import {
+  ImportedTask,
+  TaskFilters,
+  IntegrationProvider,
+  IntegrationEntry,
+  IntegrationConfig,
+  InvalidTokenError,
+  InvalidGrantError,
+} from '../integration';
+
+const authorizePath = '/oauth/authorize';
+const tokenPath = '/oauth/token';
+const redirectBaseURL = `${process.env.FRONTEND_BASE_URL}/oauth/redirect`;
+
+const oauth2 = (key: string, secret: string, host: string) =>
+  new OAuth2(key, secret, host, authorizePath, tokenPath);
+
+export const GitLabIntegration: IntegrationEntry = {
+  defaultConfig: { host: 'gitlab.com' },
+  configOptions: [
+    { field: 'projectId' },
+    { field: 'consumerkey' },
+    { field: 'privatekey', secret: true },
+  ],
+  getIntegrationProvider: (config, tokens) =>
+    new GitLabImporter(config, tokens),
+  getOAuthProvider: (config) => ({
+    instance: config.host,
+    authorizationURL: (roadmapId) =>
+      new Promise((resolve, reject) => {
+        const state = randomBytes(10).toString('hex');
+
+        const redirectUri = encodeURIComponent(
+          `${redirectBaseURL}?roadmapId=${roadmapId}&integrationName=gitLab`,
+        );
+        resolve({
+          url: `https://${config.host}${authorizePath}?client_id=${config.consumerkey}&response_type=code&state=${state}&scope=read_api&redirect_uri=${redirectUri}`,
+          token: '',
+          tokenSecret: '',
+        });
+      }),
+    swapOAuthToken: ({ verifierToken, token, tokenSecret, roadmapId }) =>
+      new Promise((resolve, reject) => {
+        const redirectUri = `${redirectBaseURL}?roadmapId=${roadmapId}&integrationName=gitLab`;
```

Comment:
This could make sense as a (top level) function
```ts
const redirectUri = (roadmapId: number) =>
  `${redirectBaseURL}?roadmapId=${roadmapId}&integrationName=gitLab`;
```